### ISSUE TITLE
%CPU numbers too large

### ISSUE THREAD

**Issue by neilhwatson** [lpefferkorn/cfe-profiler#2]
Cfe-profiler-0.1: a CFEngine profiler - http://www.loicp.eu/cfe-profiler
Looks fishy:

**\* Sorted by ticks - total CPU ticks: 140787071038223 ***
  %CPU  Namespace       Type               Bundle
13102584.00    default      agent             git_pull
13102584.00    default      agent        sync_hub_keys
13102584.00    default     common                   ps
13102584.00    default      agent             global_a
  0.01    default      agent             amd_main
  0.01    default      agent methods_loop_wrapper
  0.00    default      agent             cfbackup
  0.00    default      agent             global_s
  0.00    default     common                  tlc
  0.00    default      agent           copy_files

**Comment by lpefferkorn** [comment]
Hello Neil,

Which version of CFEngine was used ?

**Comment by neilhwatson** [comment]
Version 3.4.1.

**Comment by lpefferkorn** [comment]
@neilhwatson Could you please try again with current master ?

**Comment by neilhwatson** [comment]
Different.  This agent is configured to run in verbose mode. At the top I see this:

Cfe-profiler-0.1: a CFEngine profiler - http://www.loicp.eu/cfe-profiler

**\* Sorted by ticks - total CPU ticks: 644837548 ***
  %CPU  Namespace       Type               Bundle
 92.91    default     common                  tlc
  2.88    default     common                   ps
  1.07    default      agent            fs_inputs
  0.94    default      agent           fs_restore
  0.52    default     common               conary
  0.47    default      agent remote_bulk_template_expander
  0.40    default      agent          fs_packages
  0.31    default      agent             fs_procs
  0.15    default      agent               fsmain
  0.14    default     common                  fsg
  0.06    default  edit_line   fs_make_inputs_txt
  0.06    default      agent report_bootstrap_server
  0.04    default  edit_line   fs_expand_template
  0.03    default  edit_line      fs_insert_lines
  0.02    default  edit_line fs_append_if_no_line
cf3> SET inform = ^A
cf3> -> Bundlesequence =>  {'tlc','server','inputs','ps','amd_main','cfbackup'}

Then agent verbose output. Then at the end:

cf3> Total: Outcome of version (not specified) (agent-0): Promises observed to be kept 87%, Promises repaired 12%,
 Promises not repaired 0%
cf3>  -> Checked 2015 objects with 735 promises, i.e. model efficiency 73.27%
cf3>  -> The 735 declared promise patterns actually expanded into 8003 individual promises, i.e. declaration effic
iency 90.82%
cf3>  -> No lock purging scheduled

Cfe-profiler-0.1: a CFEngine profiler - http://www.loicp.eu/cfe-profiler

**\* Sorted by ticks - total CPU ticks: 47375428611 ***
  %CPU  Namespace       Type               Bundle
38937366528.00    default      agent             git_pull
38937366528.00    default      agent   classes_execresult
38937366528.00    default      agent disable_rhel_service
38937366528.00    default      agent        sync_hub_keys
38937366528.00    default      agent  enable_rhel_service
38937366528.00    default      agent           pam_access
38937366528.00    default      agent        ssh_host_keys
 25.13    default      agent             amd_main
 23.80    default      agent methods_loop_wrapper
 14.45    default      agent             cfbackup
  5.31    default     common                  tlc
... more like the above

**Comment by lpefferkorn** [comment]
Could you please test the debug branch, and use it this way:
  CFE_PROFILER_DEBUG=1  LD_PRELOAD=/path/to/cfe_profiler.so  cf-agent -KI | grep CFE_PROFILER_DBG

thanks.

**Comment by lpefferkorn** [comment]
Btw, is it a 32 or a 64bits system ?

**Comment by neilhwatson** [comment]
It's RHEL6 64 bit. I will not have access to that system until next week.

**Comment by neilhwatson** [comment]
Debug output:
https://gist.github.com/neilhwatson/5948250

**Comment by lpefferkorn** [comment]
Thank you Neil.

**Comment by lpefferkorn** [comment]
Ok I see, 64bits overflow

**Comment by lpefferkorn** [comment]
Neil,

Could you please try https://github.com/lpefferkorn/cfe-profiler/tree/clock_gettime branch ? 
I've tried a new measuring method, with a bonus: per bundle elapsed time in seconds instead of CPU ticks.

**Comment by neilhwatson** [comment]
Closer, but still one odd number:
Cfe-profiler-0.1: a CFEngine profiler - http://www.loicp.eu/cfe-profiler

**\* Sorted by CPU time - total CPU time: 39.26s ***
%     time(s)   Namespace     Type      Bundle
46988305367040.00 18446744027136.0 default    agent      amd_main  
29.94   11.8      default    agent      methods_loop_wrapper
16.92   6.6       default    agent      cfbackup  
10.89   4.3       default    agent      rmt_edit_template

**Comment by neilhwatson** [comment]
Note that the bundle amd_main iterates through a list of methods. It calls, directly or indirectly, almost every other bundle. Could the profiler be accumulating the run time of the these other bundles into the run time of amd_main?

**Comment by lpefferkorn** [comment]
This is still an overflow issue (the odd number 184467... looks like 2^64). Could you please try a86a229336 ? This time I have changed the way I store values.

**Comment by neilhwatson** [comment]
Looks good now. My agent is configured for verbose output.  I see multiple profiler  reports throughout. What does that mean?

**Comment by lpefferkorn** [comment]
I use this hack to trigger statistics output. I'm going to try another way...or another function really called only one time

**Comment by lpefferkorn** [comment]
@neilhwatson : Are you able to reproduce multiple reports reliably ?

**Comment by neilhwatson** [comment]
Yes, I see it consistently.

**Comment by lpefferkorn** [comment]
Could you please try the following (you will need gdb):

```
$ gdb --args  ~/tmp/cfe341/bin/cf-agent -KI
Reading symbols from /home/loic/tmp/cfe341/bin/cf-agent...done.
(gdb) b GenericDeInitialize
Breakpoint 1 at 0x406bf0
(gdb) command 1
Type commands for breakpoint(s) 1, one per line.
End with a line saying just "end".
>bt
>c
>end
(gdb) r
```

Then, for each GenericDeInitialize() call, there should be the backtrace, which is going to help me to find out why GenericDeInitialize() is called

**Comment by lpefferkorn** [comment]
@neilhwatson Could you please continue discussion about multiple reports in https://github.com/lpefferkorn/cfe-profiler/issues/5 ?

The original issue with CPU % seems fixed.

**Comment by lpefferkorn** [comment]
Fixed in branch merge 5ac8472c764fce799cc7d63df809b8c8daf1f702
